File: house-robber/samthekorean.py
# TC : O(n)
# SC : O(1)
class Solution:
    def rob(self, nums):
        n = len(nums)
        if n == 0:
            return 0
        if n == 1:
            return nums[0]

        dp = [-1] * n
        dp[0] = nums[0]
        if n > 1:
            dp[1] = max(nums[0], nums[1])

        for i in range(2, n):
            dp[i] = max(dp[i - 1], nums[i] + dp[i - 2])

        return dp[-1]

File: valid-anagram/SamTheKorean.go
// TC : O(n) : it iterates for the length of nums
// SC : O(n) : we make hashmap as the size of the given string
func isAnagram(s string, t string) bool {
    if len(s) != len(t) {
        return false
    }

    letterCounts := make(map[rune]int)

    for _, ch := range s {
        letterCounts[ch]++
    }

    for _, ch := range t {
        letterCounts[ch]--
        if letterCounts[ch] < 0 {
            return false
        }
    }

    return true
}
